### ISSUE TITLE
Windows fails to run wasm-opt.cmd, works with .exe or no extension

### ISSUE THREAD

**Issue by eugenesvk** [wasm-tool/rollup-plugin-rust#34]
On Windows you execute `wasm-opt.CMD`
https://github.com/wasm-tool/rollup-plugin-rust/blob/1f2c08b0ea6c613dd4a5f606635dcc4bab73d50f/src/index.js#L63

However, when I installed **wasm-opt** via the [binaryen](https://github.com/WebAssembly/binaryen) project the executable is `wasm-opt.EXE`
So I get the `[plugin:rust] wasm-opt failed: spawn wasm-opt.cmd ENOENT` error
If I change the line above to either `wasm-opt.EXE` or simply `wasm-opt` it works fine

Would you please add a check for an `.EXE` in case `.CMD` fails?
(by the way, does your package expects some different way of installing the wasm-opt executable with an `.cmd` extension?)

**Comment by Pauan** [comment]
You shouldn't need to install wasm-opt at all, it is automatically installed by the plugin. Are you using version 2.2.2?

**Comment by eugenesvk** [comment]
Thanks for a prompt response, yes, using the latest verion
```json
  "devDependencies": {
    "@wasm-tool/rollup-plugin-rust": "^2.2.2",
    "rimraf": "^3.0.2",
    "rollup": "^2.50.2",
    "rollup-plugin-livereload": "^2.0.0",
    "rollup-plugin-serve": "^1.1.0",
    "rollup-plugin-terser": "^7.0.2"
  },
```

But just now I've got an idea and tested with different yarn versions (all using node-modules resolution, not the newer PnP stuff, though previously I've tested with both with yarn v3) and strangely enough:

- yarn v1.22.19 does **install** `node-modules/.bin/wasm-opt.cmd` file and doesn't emit any errors (and **works**, the .wasm file is smaller)
- yarn v2.4.3 and yarn-3.2.1 also **install** `node-modules/.bin/wasm-opt.cmd` file, but DO emit the **error** mentioned above and don't work

No idea why this strange bug is happening

**Comment by Pauan** [comment]
That's incredibly strange, I wonder if it's a bug in Rollup? The plugin [catches all errors](https://github.com/wasm-tool/rollup-plugin-rust/blob/1f2c08b0ea6c613dd4a5f606635dcc4bab73d50f/src/index.js#L71-L77) from running wasm-opt, so the error is probably from wasm-opt itself.

If you enable `verbose: true` in the plugin options, what does it say?

**Comment by eugenesvk** [comment]
**TLDR**: it's yarn, the proper fix is adding binaryen as a direct dependency per https://github.com/yarnpkg/berry/issues/3109#issuecomment-880197192

I think it might be some convoluted bug in yarn resolution magic. Trying to get to the bug MVP, I've tested a simple script with `"t": "node ./test.js"` in `package.json` with only your package installed and otherwise an empty repo with 
- this command `yarn run t` and
- this simple script
```js
//test.js
const { glob, rm, mv, read, readString, exec, spawn, lock, debug } = require("./utils");
// utils.js is from your package, just commented out uninstalled/unneeded requirements there

(async () => {
  let paths = [
    "C:/test_repo/node_modules/.bin/wasm-opt.cmd",
    "wasm-opt.cmd",
  ]

  for (path of paths) {
    try      	{ await spawn(path , ["--version"], { cwd:"C:/",stdio:"inherit" });
    }catch(e)	{ console.log("fail " + path);  }
  }

})();

```


and I've noticed that it
- works with the full path and
- fails in the second `"wasm-opt.cmd"`

**However**, if I add the `"binaryen": "^106.0.0"` as a direct dependency to my package (so now I have two
```js
  "devDependencies": {
    "@wasm-tool/rollup-plugin-rust": "^2.2.2",
    "binaryen": "^106.0.0"
  },
```
)


then it's **working just fine** with both versions of the path (and if I add rollup/cargo etc. your plugin also builds just fine)
Maybe I need to use yarn link to link to binaryen explicitly (or something), though adding an explicit dependency seems like an easier workaround???

P.S.
setting `verbose: true` didn't show anything unusual, the log was identical right up to the last message `(!) Plugin rust: wasm-opt failed: spawn wasm-opt.cmd ENOENT`

**Comment by Pauan** [comment]
> the proper fix is adding binaryen as a direct dependency per https://github.com/yarnpkg/berry/issues/3109#issuecomment-880197192

But the plugin already adds binaryen as a [direct dependency](https://github.com/wasm-tool/rollup-plugin-rust/blob/1f2c08b0ea6c613dd4a5f606635dcc4bab73d50f/package.json#L31), which is why `wasm-opt.cmd` exists inside of `node_modules/.bin` in the first place.

And the plugin doesn't hardcode `node_modules/.bin` at all, it relies on Yarn resolving the PATH. So this really does seem like a bug in Yarn.

> setting verbose: true didn't show anything unusual, the log was identical right up to the last message (!) Plugin rust: wasm-opt failed: spawn wasm-opt.cmd ENOENT

Thanks for taking a look. To be clear, the `wasm-opt failed` line is a *warning*, not an error, correct? It should still build fine even if wasm-opt fails, it just means the Wasm file isn't fully optimized.

**Comment by eugenesvk** [comment]
> already adds binaryen as a [direct dependency](https://github.com/wasm-tool/rollup-plugin-rust/blob/1f2c08b0ea6c613dd4a5f606635dcc4bab73d50f/package.json#L31)

yes, I know, but the yarn team wants it as a direct dependency of **MY project**, they explicitly say they forbid relying on transitive dependencies, for example, see this https://github.com/yarnpkg/berry/issues/1875#issuecomment-698670174

>> This is intended. In npm, while you can access transitive dependencies, there is no way to be sure which ones it is, as multiple packages can depend on conflicting versions, or on no version at all.
>> Yarn aims to give you less ways to shoot yourself in the foot when it makes sense, not more, and considering that the fix on your side is simply to list a dependency you really have, it seems an acceptable trade-off.

(I'd only add that it would be an acceptable trade-off if yarn gave a clear message with a proposed fix — after all, it does see the attempt to resolve to a .cmd in a transitive dep; the way it's done now inflicts a similar pain in your foot ;) )

> doesn't hardcode node_modules/.bin at all

the yarn team approves :) as this folder is an "implementation detail", but that's a separate thing from direct vs transitive deps

> line is a _warning_, not an error, correct

yes, built fine, wasm file was bigger

**Comment by Pauan** [comment]
> yes, I know, but the yarn team wants it as a direct dependency of MY project, they explicitly say they forbid relying on transitive dependencies

Ah, I see, that's a huge pain... that means that everybody who wants to use this plugin has to add `binaryen` as a dependency.

I can understand their argument, but it *really* hurts things like plugins... I get the impression they haven't thought about plugins, and their restriction was intended for apps.

Maybe I can workaround it by specifying `binaryen` as a `peerDependency`? That would at least make it clear that the user has to install binaryen themself.

> yes, built fine, wasm file was bigger

That's good to know, so at least it doesn't completely break things, but having unoptimized files is still not great.

**Comment by eugenesvk** [comment]
> I can understand their argument, but it really hurts things like plugins... I get the impression they haven't thought about plugins, and their restriction was intended for apps.

Yeah, to me it also seems like the current architecture is worse, I'd much prefer a warning + giving priority to direct dependencies rather than outright blocking them:
  - your package would work without an explicit dependency on binaryen in my package
  - if there is another package that depends on another version of binaryen, then I'd get a warning that there is a potential conflict that I might need to resolve by adding an explicit dependency (or continue to ignore if it doesn't break anything)

... or something like that

> Maybe I can workaround it by specifying `binaryen` as a `peerDependency`? That would at least make it clear that the user has to install binaryen themself.

Good idea, this would give that currently missing warning (though slightly burden everyone with other build tools that aren't that strict)!
